[docs] Prepare message docs for sync to docs.microsoft.com

brendanzagaeski opened the pull request:
I have a couple questions to ask before these changes will be ready to merge:

1. @conceptdev, is it OK that the metadata in this commit only includes `title` and `ms.date` fields? Did you have other fields in mind to include on these pages? I noticed that the contribution guidelines say the `ms.assetid` is no longer required, so I tentatively omitted that for now. But let me know if I should add `ms.assetid` or any another fields.
2. @conceptdev, @jonpryor, and anyone else on the team who's interested, I think it might work well to update these pages to match the format of the [NuGet error reference](https://docs.microsoft.com/nuget/reference/errors-and-warnings/nu1103) pages. I like the way that format looks. Each page starts with a code block (a pre-formatted section) that shows the error message, followed by an **Issue** section, and then a **Solution** section. There are also examples for how to handle [multiple Issues and Solutions for one error code](https://docs.microsoft.com/nuget/reference/errors-and-warnings/nu3019). Does that look good to you all too? If so, I'll go ahead with reformatting these pages to match.

Thanks!

jonpryor commented:
>  I think it might work well to update these pages to match the format of the NuGet error reference pages.

I agree.  If we squint *just right*, that's also how the CSC error pages are, e.g. [CS0015](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0015), in which the *first* paragraph is the error message produced.

The NuGet version is nicer in that the error message is more clearly delineated as an error message.

Either way, having the error message *first* is a good idea.

jonpryor commented on Documentation/guides/messages/adb0000.md:
```diff
@@ -1,20 +1,24 @@
+title: Compiler error/warning ADB0000
+ms.date: 07/04/2018
+---
+# Compiler error/warning ADB0000
```
I don't understand how/why this line changed.  Byte-order mark?

brendanzagaeski replied:
There were a few cases of byte-order marks that changed. In this particular case though, the changes was just from lower-casing "Error" and "Warning" to follow [sentence-style capitalization](https://github.com/MicrosoftDocs/xamarin-docs/blob/live/contributing-guidelines/template.md#headings). I debated on this change because the C# compiler error reference pages and NuGet error reference pages do capitalize "Error" and "Warning" in the titles. I'll be sure to get Craig's input about which option to choose before this PR is finalized.

On a separate question, is it OK to drop the byte-order marks from files that have them? The contribution guidelines don't specifically mention file encoding or byte-order marks. I spot-checked a few files in the C# compiler error reference pages and NuGet error reference pages and didn't see byte-order marks in those few, so it seems maybe it's most common not to include a byte-order marks. For now, I'll tentatively continue to leave out the byte-order marks for the next draft.

conceptdev replied:
BOM is not required - I would remove them if you notice them

jonpryor commented on Documentation/guides/messages/adb0000.md:
```diff
@@ -1,20 +1,24 @@
-SDK and is used internally by Xamarin.Android to communicate with
-Android emulators and devices. Learn more about `adb` from the
-[Android documentation][adb].
+unhandled/unexpected error or warning.  `adb` is part of the Android SDK and is
```
Do we need to "re-format" *everything*?  The only changes here are "whitespace" changes -- wider line lengths, two spaces after a `.`, etc.  No "actual" content has changed.  (Some minor formatting changes in subsequent paragraphs.)

Is it necessary to do all this?

Was this reformatting done manually or via tooling?

brendanzagaeski replied:
This reformatting was done manually, but I didn't have a strong opinion that motivated it, so in the next draft I'll plan to go back to using the original spacing after `.` for each document and only rewrap a paragraph if the line length has changed.

conceptdev commented:
Love it in principle - been a busy week with //build so can I take another day to take a closer look?

My initial feedback is that we should choose a different folder name for SEO. One of the links you shared uses `errors-warnings` which i think captures the two most likely terms customers would use? Happy to discuss.

RE: your question, `description:` is also required by the tooling (don't ask me why...). Can we pull something from the content into the YAML for the description? All other fields that are "required", (author, ms.author) we can set globally in a config file, so you do **not** need to repeat them in every page.

brendanzagaeski commented:
> Love it in principle - been a busy week with //build so can I take another day to take a closer look?

Yup, no problem! I'll plan to add the candidate change to arrange the documents in the NuGet error reference format in a separate commit so you can look at this original option and compare it to that one if you like.

brendanzagaeski commented:
I have updated the pull request with a second draft. The original draft is still included as the first commit, but the second commit now starts over from how the files were originally. I have split several of the changes into separate commits this time so they can be examined and considered separately if desired.

Changes *added* in the second draft:

  * Added a `description:` metadata field to all the pages. I borrowed the style of `description:` field from the [NuGet error reference pages](https://github.com/NuGet/docs.microsoft.com-nuget/blob/master/docs/reference/errors-and-warnings/NU1103.md).
  * Reformatted the body of the pages to follow the NuGet error reference pages. I made one addition to that format to include an "Example messages" header at the top of pages. That will hopefully help avoid potential confusion for cases where multiple different messages are listed or where the exact text of the message depends on names from the user's project.
  * Updated some outdated links and changed a couple relative links to full-length links.
  * In the last commit, renamed "Compiler" to "Xamarin.Android" in the titles of the pages. The motivation was that the NuGet error reference pages say "NuGet" rather than "Compiler", and I think that approach might work well for Xamarin.Android too, especially since some of the errors are deployment errors rather than build errors. But I'm totally happy to undo that if it's preferable to stay with "Compiler."

Changes *no longer included* in the second draft:

  * No new spaces added between existing sentences
  * No paragraph rewrapping except for a few cases where sentences were split into their own new paragraphs
  * No changes to list indentation or bullet characters
  * No addition of `<PropertyGroup/>` elements around XML MSBuild property examples (I'll tentatively plan to make a separate follow-up PR to add these just in case it might be nice to have them.)
  * No added newlines at the ends of files that didn't have them already

Potential other items to change:

  * I have not yet renamed the parent directory. I think the current name `messages` might have been originally inspired by the [C# error reference pages](https://docs.microsoft.com/dotnet/csharp/language-reference/compiler-messages/cs0001). But `errors-and-warnings` does sound good to me. We could change it here or just in the (not yet created) xamarin-engineering-docs-pr mirror, whichever seems best.

  * I escaped the `\` path separator character as `\\`. Technically this isn't necessary in GitHub Flavored Markdown unless the next character after `\` is a punctuation character. I could remove the extra escapes where they aren't needed.

  * Some of the pages recommend reporting an issue. Some link to <https://github.com/xamarin/xamarin-android/wiki/Submitting-Bugs,-Feature-Requests,-and-Pull-Requests> and say:

    > Consider submitting a [bug](https://github.com/xamarin/xamarin-android/wiki/Submitting-Bugs,-Feature-Requests,-and-Pull-Requests) if you are getting this warning under normal circumstances.

    A few instead link to <https://github.com/xamarin/xamarin-android/issues/new> and say:

    > If the problem persists, raise an [issue](https://github.com/xamarin/xamarin-android/issues/new). 

    When these pages are shown on docs.microsoft.com, I think they will all automatically get a **Feedback** section:

    ![image](https://user-images.githubusercontent.com/4934473/57578685-df558500-7445-11e9-88ac-3ea1b1a7428f.png)

    Maybe it would be appropriate to adjust the wording for these sentences to point users to the **This product** button under the **Feedback** section on these pages? Or do we also want the wording to make sense when users are browsing through the files directly in the xamarin-android repo?

  * Several of the pages include a sentence similar to:

    > Note that nothing in the open source xamarin-android repository emits ADB0000, as features such as debugging and "fast deployment" are implemented in the proprietary Xamarin.Android additions.

    It crossed my mind that this might look out of place when the pages are on docs.microsoft.com. It is useful information for users who usually try to dig into the Xamarin.Android source code to understand errors they are seeing, but in these cases it's actually saying that the source code is *not* available, so it might be a little noisy for most users. I'll brainstorm to see if I have any ideas for a small rewording that might smooth that out.

  * For a tiny extra bit of consistency, I could optionally adjust all the example messages so that none include the error code prefix (e.g., `error XA0000:`). It looks like that's the convention the NuGet and C# error reference pages follow, but this is quite a small consideration, so it seems fine to leave them as they are too.

jonpryor commented:
> It crossed my mind that this might look out of place when the pages are on docs.microsoft.com.

It might look out of place, but it's also small/not terribly voluminous, so I don't think we need to remove this comment.  We might want to consider having this as a "note" of some form at the bottom, so that "more useful" information is above it.

jonpryor commented on Documentation/guides/messages/adb0000.md:
```diff
@@ -1,20 +1,33 @@
-Note that nothing in the open source `xamarin-android` repository
-emits `ADB0000`, as features such as debugging and "fast deployment"
-are implemented in the proprietary Xamarin.Android bits.
+Note that nothing in the open source xamarin-android repository
```
...e.g. we should consider moving this *after* the `## Solution` section.  Perhaps as an `## Implementation Notes` section?

@conceptdev: Is there any guidance on providing "hints regarding source code"?

conceptdev replied:
I don't have a strong preference, but I like the idea of a separate heading (at the end)... this would improve readability, since it wouldn't distract from the solution.

conceptdev commented:
FYI I have created a test branch to preview these on docs

https://review.docs.microsoft.com/en-us/xamarin/android/warnings-errors/?branch=android-error-messages

brendanzagaeski commented:
Changes in the latest force-push:

  * Moved the "nothing in the open source xamarin-android repository" paragraphs under a new "Implementation notes" header.
  * Added a few links that were missing from the individual commit messages for the relevant sections of the documentation style guidelines.

Not changed:

  * I didn't change the wording related to bug reporting yet, but I think that is OK leave as-is for this pull request. We can wait to see how the pages look on docs.microsoft.com and decide if it makes sense to adjust that wording in a future pull request.
  * We can also follow-up in a separate pull request as needed if we decide to rename the directory or include a table of contents and index page here in the xamarin-android repo. My current idea is that I can expand <https://github.com/xamarin/xamarin-android/pull/2789> to include a step in the Makefile targets to generate the table of contents and index page.

This takes care of the review comments and the questions I had, so I think it's all set to merge now. Thanks for the discussion and sanity-checking!

conceptdev commented:
@brendanzagaeski 

* xa0115.md - there are two image tags, we need the image file included
* xa0100.md and xa0104.md are missing in the sequence - is this intentional?

brendanzagaeski commented:
@conceptdev, ooh yes, apologies. I meant to follow up on the email thread about those questions.

> xa0115.md - there are two image tags, we need the image files included

The images are currently in <https://github.com/xamarin/xamarin-android/tree/master/Documentation/images>. But thinking ahead about automating the synchronization, I think it probably makes sense to move the images for these pages to a subdirectory alongside the Markdown files, so I've set up another PR for that: <https://github.com/xamarin/xamarin-android/pull/3151>.

> xa0100.md and xa0104.md are missing in the sequence - is this intentional?

I think it's "sort of" intentional, in that I suspect no pages were ever written for `xa0100.md` or `xa0104.md`, but they were added to the list of error messages to make the list more complete (even in the absence of dedicated pages for them). I think it might be OK to leave them as unlinked items in the list for now, and then add links if pages are written for them in the future. I've included a commit to remove those links as part of <https://github.com/xamarin/xamarin-android/pull/3151> too.
